From ee65698463032973085718f8f1ff07a62bb94456 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 26 Aug 2014 14:11:55 -0400
Subject: [ethernet] igb: Fix memory leak in igb_get_module_eeprom()

Message-id: <1409062359-13063-38-git-send-email-sassmann@redhat.com>
Patchwork-id: 89153
O-Subject: [RHEL7.1 PATCH 37/81] igb: Fix memory leak in igb_get_module_eeprom()
Bugzilla: 1091121
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Christian Engelmayer <cengelma@gmx.at>

Fix a memory leak in the igb_get_module_eeprom() error handling path.
Detected by Coverity: CID 1016508.

Signed-off-by: Christian Engelmayer <cengelma@gmx.at>
Tested-by: Jeff Pieper <jeffrey.e.pieper@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit db41b87db93e1ef28b6e9c4035f1cdc17e2eceb2)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/igb/igb_ethtool.c b/drivers/net/ethernet/intel/igb/igb_ethtool.c
index 170e4db..d668815 100644
--- a/drivers/net/ethernet/intel/igb/igb_ethtool.c
+++ b/drivers/net/ethernet/intel/igb/igb_ethtool.c
@@ -2791,9 +2791,11 @@ static int igb_get_module_eeprom(struct net_device *netdev,
  /* Read EEPROM block, SFF-8079/SFF-8472, word at a time */
  for (i = 0; i < last_word - first_word + 1; i++) {
   status = igb_read_phy_reg_i2c(hw, first_word + i, &dataword[i]);
-  if (status != E1000_SUCCESS)
+  if (status != E1000_SUCCESS) {
    /* Error occurred while reading module */
+   kfree(dataword);
    return -EIO;
+  }
 
   be16_to_cpus(&dataword[i]);
  }
-- 
1.7.1